### PR TITLE
Add milestone shares scores to MilestonesEditor

### PR DESCRIPTION
**Pull request by IidaKainu**
Calculate the difference between the actual and target shares. If the result is more than a 30% decrease, it is considered as a poor score.

### CODE REVIEW

File: frontend/src/components/TaskValueCreatedVisualization.tsx [hunk 1]
```diff
@@ -14,23 +14,37 @@ import { apiV2 } from '../api/api';
 
 const classes = classNames.bind(css);
 
+// either provide versions or customer stakes
 export const TaskValueCreatedVisualization: FC<{
-  versions: Version[];
+  stakes?: CustomerStakes[];
+  versions?: Version[];
```
Comment (juhoeerola) [inline frontend/src/components/TaskValueCreatedVisualization.tsx:20]:
```suggestion
  ({ stakes: CustomerStakes[] } | { versions: Version[] }) & {
```

The "either" could be expressed in the type, obsoleting the comment

File: frontend/src/components/TaskValueCreatedVisualization.tsx [hunk 2]
```diff
@@ -14,23 +14,37 @@ import { apiV2 } from '../api/api';
...
-}> = ({ versions, width, height, noTooltip, vertical, barWidth = 37 }) => {
+}> = ({
+  stakes,
+  versions,
+  width,
+  height,
+  noTooltip,
+  vertical,
+  barWidth = 37,
+}) => {
```
Comment (juhoeerola) [inline frontend/src/components/TaskValueCreatedVisualization.tsx:34]:
```suggestion
}> = ({ width, height, noTooltip, vertical, barWidth = 37, ...props }) => {
  const { stakes, versions } = {
    stakes: undefined,
    versions: undefined,
    ...props,
  };
```

The provided option can then be unpacked providing `undefined` for the other

File: frontend/src/utils/CustomerUtils.ts [hunk 3]
```diff
@@ -25,3 +25,18 @@ export const randomColor = (
...
+    .sort((a, b) => b.weight - a.weight)
+    .map(({ id, weight }) => ({
+      id,
+      percent: weight / totalValue,
+    }));
+};
+
+export const stakesDifferTooMuchFromPlanned = (
+  decimal: number,
+  targetDecimal: number,
```
Comment (juhoeerola) [inline frontend/src/utils/CustomerUtils.ts:41]: The parameter names could be a bit more descriptive e.g.
```suggestion
  actual: number,
  planned: number,
```

File: frontend/src/components/CustomerStakesScore.tsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+}> = ({ version, completed }) => {
+  const roadmapId = useSelector(chosenRoadmapIdSelector);
+  const { data: customers } = apiV2.useGetCustomersQuery(
+    roadmapId ?? skipToken,
+  );
+  const [targetData, setTargetData] = useState<TargetStakes[] | undefined>(
+    undefined,
+  );
+  const [data, setData] = useState<CustomerStakes[]>([]);
+  const [totalValue, setTotalValue] = useState<number | undefined>(undefined);
```
Comment (juhoeerola) [inline frontend/src/components/CustomerStakesScore.tsx:38]:
```suggestion
  const [totalValue, setTotalValue] = useState(0);
```
This is only read once using non-null assertion, so a default value could be used here instead

File: frontend/src/components/CustomerStakesScore.tsx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+  useEffect(() => {
+    if (!targetData) return;
+
+    const customerStakes = Array.from(
+      totalCustomerStakes(version.tasks, customers),
+    );
+    const stakesTotalValue = customerStakes.reduce(
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-unused-vars
+      (acc, [_, value]) => acc + value,
```
Comment (juhoeerola) [inline frontend/src/components/CustomerStakesScore.tsx:52]:
```suggestion
      (acc, [, value]) => acc + value,
```
The name can just be omitted

File: frontend/src/components/CustomerStakesScore.tsx [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+      <div>
+        <div
+          className={classes(css.score, {
+            [css.completed]: completed,
+            [css.differsFromPlanned]: scoresDifferFromPlanned,
+          })}
+        >
+          <Trans i18nKey="Client shares score" />
+          <div className={classes(css.result)}>
+            <Trans i18nKey={scoresDifferFromPlanned ? 'Poor' : 'Great'} />
```
Comment (juhoeerola) [inline frontend/src/components/CustomerStakesScore.tsx:117]: These keys seem to be missing from translations

**Review by juhoeerola — CHANGES_REQUESTED** [review]

**Comment by IidaKainu** [comment]
Thanks Juho, the change requests were absolutely great. Added changes (and now I'm motivated to improve my typescript knowledge)

File: frontend/src/components/CustomerStakesScore.tsx (hunk 6 above)
Comment (IidaKainu) [reply frontend/src/components/CustomerStakesScore.tsx:115]: Totally missed that one. Added

**Review by juhoeerola — APPROVED** [review]

File: frontend/src/utils/CustomerUtils.ts [hunk 7]
```diff
@@ -25,3 +25,18 @@ export const randomColor = (
...
+    .sort((a, b) => b.weight - a.weight)
+    .map(({ id, weight }) => ({
+      id,
+      percent: weight / totalValue,
+    }));
+};
+
+export const stakesDifferTooMuchFromPlanned = (
+  actualDecimal: number,
+  plannedDecimal: number,
```
[the remark this answers is not in the document]
Comment (IidaKainu) [reply frontend/src/utils/CustomerUtils.ts]: Yeah it has been a while for me working with math: it does not matter. Changed and unified the other variable names too

**Pull request merged**